### ISSUE TITLE
If a `NetworkImage` fails due to transient network issues, it's difficult for applications to catch and retry that.

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#110359]
Specifically, we make sure to evict from the cache here: https://github.com/flutter/flutter/blob/000b96cfdf848d5accb2da8fe455fb19d01e3b9c/packages/flutter/lib/src/painting/_network_image_io.dart#L141-L146

This makes sure that if you do retry, you won't just keep getting the stream that's in an error state.

However, it's relatively challenging for an application to handle this using the `onError` callback for `Image` or `precacheImage`. This is likely a cause for bugs like https://github.com/flutter/flutter/issues/64558.

It would probably make sense to have something like a `RetryImageProvider` that could be used similarly to `ResizeImageProvider` and would retry byte acquisition on failure. I don't have a strong opinion about whether this belongs in the framework or a separate package.

@jonahwilliams @goderbauer might have good ideas too.

**Comment by dnfield** [comment]
I'm a bit torn on whether this is a P5 - new feature, or a P4. The bug that this causes is flakey and difficult to diagnose.

**Comment by jonahwilliams** [comment]
I'll state for the record (since I've already said it privately) that we should not try to make the existing providers any smarter.

**Comment by dnfield** [comment]
I agree. I think a composition could work well here though with something like `RetryImage` that would be used to wrap network-based images (or perhaps a file based image being fetch from a network share).
